Cannot build

I want to make a project on v0.3.3, but build fails:

```
Compiling psp v0.3.3
error[E0605]: non-primitive cast: `*mut &mut dyn BoxMeUp` as `&mut dyn BoxMeUp`
   --> /home/slendi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/psp-0.3.3/src/panic.rs:151:42
    |
151 |         panic_unwind::__rust_start_panic(obj as _)
    |                                          ^^^^^^^^ invalid cast
    |
help: consider borrowing the value
    |
151 |         panic_unwind::__rust_start_panic(&mut *obj as _)
    |                                          ++++++

error[E0277]: the trait bound `&mut dyn BoxMeUp: BoxMeUp` is not satisfied
   --> /home/slendi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/psp-0.3.3/src/panic.rs:151:42
    |
151 |         panic_unwind::__rust_start_panic(obj as _)
    |                                          ^^^ the trait `BoxMeUp` is not implemented for `&mut dyn BoxMeUp`
    |
    = help: the trait `BoxMeUp` is implemented for `PanicPayload<'a>`
    = note: required for the cast from `&mut dyn BoxMeUp` to the object type `dyn BoxMeUp`

Some errors have detailed explanations: E0277, E0605.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `psp` (lib) due to 2 previous errors
```

Same error also happens on v0.3.2.

--- a/README.md
+++ b/README.md
@@ -56,7 +56,7 @@ implementations of graphics functions, and the addition of missing libraries.
 ## Dependencies
 
 To compile for the PSP, you will need a Rust **nightly** version equal to or
-later than `2023-04-11` and the `rust-src` component. Please install Rust using
+later than `2023-03-28` and the `rust-src` component. Please install Rust using
 https://rustup.rs/
 
 Use the following if you are new to Rust. (Feel free to set an override manually

--- a/README.md
+++ b/README.md
@@ -56,7 +56,7 @@ implementations of graphics functions, and the addition of missing libraries.
 ## Dependencies
 
 To compile for the PSP, you will need a Rust **nightly** version equal to or
-later than `2022-09-04` and the `rust-src` component. Please install Rust using
+later than `2023-04-11` and the `rust-src` component. Please install Rust using
 https://rustup.rs/
 
 Use the following if you are new to Rust. (Feel free to set an override manually

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -150,7 +150,7 @@ fn main() {
     }
 
     let old_version = MINIMUM_RUSTC_VERSION
-        >= Version {
+        > Version {
             // Remove `-nightly` pre-release tag for comparison.
             pre: Vec::new(),
             ..rustc_version.semver.clone()
@@ -160,7 +160,7 @@ fn main() {
         None => false,
         Some(date) => {
             MINIMUM_COMMIT_DATE
-                >= CommitDate::parse(&date).expect("could not parse `rustc --version` commit date")
+                > CommitDate::parse(&date).expect("could not parse `rustc --version` commit date")
         }
     };
 

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -140,7 +140,7 @@ const MINIMUM_RUSTC_VERSION: Version = Version {
 fn main() {
     let rustc_version = rustc_version::version_meta().unwrap();
 
-    if rustc_version.channel >= Channel::Nightly {
+    if rustc_version.channel > Channel::Nightly {
         println!("cargo-psp requires a nightly rustc version.");
         println!(
             "Please run `rustup override set nightly` to use nightly in the \

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -143,7 +143,7 @@ const MINIMUM_COMMIT_DATE: CommitDate = CommitDate {
 };
 const MINIMUM_RUSTC_VERSION: Version = Version {
     major: 1,
-    minor: 69,
+    minor: 70,
     patch: 0,
     pre: Vec::new(),
     build: Vec::new(),

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -121,17 +121,17 @@ impl fmt::Display for CommitDate {
     }
 }
 
-// Minimum 2022-09-03, remember to update both commit date and version too,
+// Minimum 2023-04-11, remember to update both commit date and version too,
 // below. Note that the `day` field lags by one day, as the toolchain always
 // contains the previous days' nightly rustc.
 const MINIMUM_COMMIT_DATE: CommitDate = CommitDate {
-    year: 2022,
-    month: 09,
-    day: 03,
+    year: 2023,
+    month: 04,
+    day: 11,
 };
 const MINIMUM_RUSTC_VERSION: Version = Version {
     major: 1,
-    minor: 63,
+    minor: 70,
     patch: 0,
     pre: Vec::new(),
     build: Vec::new(),

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -140,7 +140,7 @@ const MINIMUM_RUSTC_VERSION: Version = Version {
 fn main() {
     let rustc_version = rustc_version::version_meta().unwrap();
 
-    if rustc_version.channel > Channel::Nightly {
+    if rustc_version.channel >= Channel::Nightly {
         println!("cargo-psp requires a nightly rustc version.");
         println!(
             "Please run `rustup override set nightly` to use nightly in the \
@@ -150,7 +150,7 @@ fn main() {
     }
 
     let old_version = MINIMUM_RUSTC_VERSION
-        > Version {
+        >= Version {
             // Remove `-nightly` pre-release tag for comparison.
             pre: Vec::new(),
             ..rustc_version.semver.clone()
@@ -160,7 +160,7 @@ fn main() {
         None => false,
         Some(date) => {
             MINIMUM_COMMIT_DATE
-                > CommitDate::parse(&date).expect("could not parse `rustc --version` commit date")
+                >= CommitDate::parse(&date).expect("could not parse `rustc --version` commit date")
         }
     };
 

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -121,17 +121,29 @@ impl fmt::Display for CommitDate {
     }
 }
 
-// Minimum 2023-04-11, remember to update both commit date and version too,
+impl core::ops::Add<CommitDate> for CommitDate {
+    type Output = CommitDate;
+
+    fn add(self, rhs: CommitDate) -> Self::Output {
+        Self {
+            year: self.year + rhs.year,
+            month: self.month + rhs.month,
+            day: self.day + rhs.day,
+        }
+    }
+}
+
+// Minimum 2023-03-27, remember to update both commit date and version too,
 // below. Note that the `day` field lags by one day, as the toolchain always
 // contains the previous days' nightly rustc.
 const MINIMUM_COMMIT_DATE: CommitDate = CommitDate {
     year: 2023,
-    month: 04,
-    day: 11,
+    month: 03,
+    day: 27,
 };
 const MINIMUM_RUSTC_VERSION: Version = Version {
     major: 1,
-    minor: 70,
+    minor: 69,
     patch: 0,
     pre: Vec::new(),
     build: Vec::new(),
@@ -167,7 +179,12 @@ fn main() {
     if old_version || old_commit {
         println!(
             "cargo-psp requires rustc nightly version >= {}",
-            MINIMUM_COMMIT_DATE,
+            MINIMUM_COMMIT_DATE + 
+            CommitDate {
+                year: 0,
+                month: 0,
+                day: 1
+            },
         );
         println!("Please run `rustup update nightly` to upgrade your nightly version");
 

--- a/cargo-psp/src/main.rs
+++ b/cargo-psp/src/main.rs
@@ -179,12 +179,12 @@ fn main() {
     if old_version || old_commit {
         println!(
             "cargo-psp requires rustc nightly version >= {}",
-            MINIMUM_COMMIT_DATE + 
-            CommitDate {
-                year: 0,
-                month: 0,
-                day: 1
-            },
+            MINIMUM_COMMIT_DATE
+                + CommitDate {
+                    year: 0,
+                    month: 0,
+                    day: 1
+                },
         );
         println!("Please run `rustup update nightly` to upgrade your nightly version");
 

--- a/ci/concourse/build-rust-macos.yml
+++ b/ci/concourse/build-rust-macos.yml
@@ -1,7 +1,7 @@
 platform: darwin
 
 params:
-  RUSTUP_TOOLCHAIN: nightly-2023-04-11
+  RUSTUP_TOOLCHAIN: nightly-2023-03-28
 
 inputs:
   - name: repo

--- a/ci/concourse/build-rust-macos.yml
+++ b/ci/concourse/build-rust-macos.yml
@@ -1,7 +1,7 @@
 platform: darwin
 
 params:
-  RUSTUP_TOOLCHAIN: nightly-2022-09-04
+  RUSTUP_TOOLCHAIN: nightly-2023-04-11
 
 inputs:
   - name: repo

--- a/ci/concourse/build-rust.yml
+++ b/ci/concourse/build-rust.yml
@@ -11,7 +11,7 @@ image_resource:
     tag: 1.44-slim
 
 params:
-  RUSTUP_TOOLCHAIN: nightly-2023-04-11
+  RUSTUP_TOOLCHAIN: nightly-2023-03-28
 
 inputs:
   - name: repo

--- a/ci/concourse/build-rust.yml
+++ b/ci/concourse/build-rust.yml
@@ -11,7 +11,7 @@ image_resource:
     tag: 1.44-slim
 
 params:
-  RUSTUP_TOOLCHAIN: nightly-2022-09-04
+  RUSTUP_TOOLCHAIN: nightly-2023-04-11
 
 inputs:
   - name: repo

--- a/psp/src/panic.rs
+++ b/psp/src/panic.rs
@@ -145,9 +145,9 @@ extern "C-unwind" {
 #[inline(never)]
 #[no_mangle]
 #[cfg(not(feature = "std"))]
-fn rust_panic(mut msg: &mut dyn BoxMeUp) -> ! {
+fn rust_panic(msg: &mut dyn BoxMeUp) -> ! {
     let code = unsafe {
-        let obj = &mut msg as *mut &mut dyn BoxMeUp;
+        let obj = msg;
         panic_unwind::__rust_start_panic(obj as _)
     };
 

--- a/psp/src/sys/hprm.rs
+++ b/psp/src/sys/hprm.rs
@@ -2,7 +2,7 @@
 
 bitflags::bitflags! {
     #[repr(transparent)]
-    pub struct Key: u32 {
+    pub struct HprmKey: u32 {
         const PLAY_PAUSE  = 0x1;
         const FORWARD     = 0x4;
         const BACK        = 0x8;
@@ -27,7 +27,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// < 0 on error
-    pub fn sceHprmPeekCurrentKey(key: *mut Key) -> i32;
+    pub fn sceHprmPeekCurrentKey(key: *mut HprmKey) -> i32;
 
     #[psp(0x2BCEC83E)]
     /// Peek at the current latch data.

--- a/psp/src/sys/mp3.rs
+++ b/psp/src/sys/mp3.rs
@@ -24,7 +24,7 @@ pub struct SceMp3InitArg {
 
 #[derive(Copy, Clone, Debug)]
 #[repr(transparent)]
-pub struct Handle(pub i32);
+pub struct Mp3Handle(pub i32);
 
 psp_extern! {
     #![name = "sceMp3"]
@@ -51,7 +51,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 if success, < 0 on error.
-    pub fn sceMp3ReleaseMp3Handle(handle: Handle) -> i32;
+    pub fn sceMp3ReleaseMp3Handle(handle: Mp3Handle) -> i32;
 
     #[psp(0x35750070)]
     /// # Return Value
@@ -73,7 +73,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 if success, < 0 on error.
-    pub fn sceMp3Init(handle: Handle) -> i32;
+    pub fn sceMp3Init(handle: Mp3Handle) -> i32;
 
     #[psp(0xD021C0FB)]
     /// # Parameters
@@ -84,7 +84,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// number of bytes in decoded pcm buffer, < 0 on error.
-    pub fn sceMp3Decode(handle: Handle, dst: *mut *mut i16) -> i32;
+    pub fn sceMp3Decode(handle: Mp3Handle, dst: *mut *mut i16) -> i32;
 
     #[psp(0xA703FE0F)]
     /// # Parameters
@@ -98,7 +98,7 @@ psp_extern! {
     ///
     /// 0 if success, < 0 on error.
     pub fn sceMp3GetInfoToAddStreamData(
-        handle: Handle,
+        handle: Mp3Handle,
         dst: *mut *mut u8,
         to_write: *mut i32,
         src_pos: *mut i32,
@@ -113,7 +113,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 if success, < 0 on error.
-    pub fn sceMp3NotifyAddStreamData(handle: Handle, size: i32) -> i32;
+    pub fn sceMp3NotifyAddStreamData(handle: Mp3Handle, size: i32) -> i32;
 
     #[psp(0xD0A56296)]
     /// # Parameters
@@ -123,7 +123,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 1 if more stream data is needed, < 0 on error.
-    pub fn sceMp3CheckStreamDataNeeded(handle: Handle) -> i32;
+    pub fn sceMp3CheckStreamDataNeeded(handle: Mp3Handle) -> i32;
 
     #[psp(0x3CEF484F)]
     /// # Parameters
@@ -134,7 +134,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 if success, < 0 on error.
-    pub fn sceMp3SetLoopNum(handle: Handle, loop_: i32) -> i32;
+    pub fn sceMp3SetLoopNum(handle: Mp3Handle, loop_: i32) -> i32;
 
     #[psp(0xD8F54A51)]
     /// # Parameters
@@ -144,7 +144,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// Number of loops
-    pub fn sceMp3GetLoopNum(handle: Handle) -> i32;
+    pub fn sceMp3GetLoopNum(handle: Mp3Handle) -> i32;
 
     #[psp(0x354D27EA)]
     /// # Parameters
@@ -154,7 +154,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// Number of decoded samples
-    pub fn sceMp3GetSumDecodedSample(handle: Handle) -> i32;
+    pub fn sceMp3GetSumDecodedSample(handle: Mp3Handle) -> i32;
 
     #[psp(0x87C263D1)]
     /// # Parameters
@@ -164,7 +164,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// Number of max samples to output
-    pub fn sceMp3GetMaxOutputSample(handle: Handle) -> i32;
+    pub fn sceMp3GetMaxOutputSample(handle: Mp3Handle) -> i32;
 
     #[psp(0x8F450998)]
     /// # Parameters
@@ -174,7 +174,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// Sampling rate of the mp3
-    pub fn sceMp3GetSamplingRate(handle: Handle) -> i32;
+    pub fn sceMp3GetSamplingRate(handle: Mp3Handle) -> i32;
 
     #[psp(0x87677E40)]
     /// # Parameters
@@ -184,7 +184,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// Bitrate of the mp3
-    pub fn sceMp3GetBitRate(handle: Handle) -> i32;
+    pub fn sceMp3GetBitRate(handle: Mp3Handle) -> i32;
 
     #[psp(0x7F696782)]
     /// # Parameters
@@ -194,7 +194,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// Number of channels of the mp3
-    pub fn sceMp3GetMp3ChannelNum(handle: Handle) -> i32;
+    pub fn sceMp3GetMp3ChannelNum(handle: Mp3Handle) -> i32;
 
     #[psp(0x2A368661)]
     /// # Parameters
@@ -204,5 +204,5 @@ psp_extern! {
     /// # Return Value
     ///
     /// < 0 on error
-    pub fn sceMp3ResetPlayPosition(handle: Handle) -> i32;
+    pub fn sceMp3ResetPlayPosition(handle: Mp3Handle) -> i32;
 }

--- a/psp/src/sys/registry.rs
+++ b/psp/src/sys/registry.rs
@@ -6,12 +6,12 @@ pub const REG_KEYNAME_SIZE: u32 = 27;
 
 /// Typedef for a registry handle.
 #[repr(transparent)]
-pub struct Handle(u32);
+pub struct RegistryHandle(u32);
 
 /// Struct used to open a registry.
 #[repr(C)]
 #[derive(Copy, Clone)]
-pub struct Key {
+pub struct RegistryKey {
     pub key_type: KeyType,
     /// Seemingly never used, set to `SYSTEM_REGISTRY`.
     pub name: [u8; 256usize],
@@ -54,9 +54,9 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegOpenRegistry(
-        reg: *mut Key,
+        reg: *mut RegistryKey,
         mode: i32,
-        handle: *mut Handle,
+        handle: *mut RegistryHandle,
     ) -> i32;
 
     #[psp(0x39461B4D)]
@@ -69,7 +69,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 on success, < 0 on error
-    pub fn sceRegFlushRegistry(handle: Handle) -> i32;
+    pub fn sceRegFlushRegistry(handle: RegistryHandle) -> i32;
 
     #[psp(0xFA8A5739)]
     /// Close the registry
@@ -81,7 +81,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 on success, < 0 on error
-    pub fn sceRegCloseRegistry(handle: Handle) -> i32;
+    pub fn sceRegCloseRegistry(handle: RegistryHandle) -> i32;
 
     #[psp(0x1D8A762E)]
     /// Open a registry directory
@@ -97,10 +97,10 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegOpenCategory(
-        handle: Handle,
+        handle: RegistryHandle,
         name: *const u8,
         mode: i32,
-        dir_handle: *mut Handle,
+        dir_handle: *mut RegistryHandle,
     ) -> i32;
 
     #[psp(0x4CA16893)]
@@ -115,7 +115,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegRemoveCategory(
-        handle: Handle,
+        handle: RegistryHandle,
         name: *const u8,
     ) -> i32;
 
@@ -129,7 +129,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 on success, < 0 on error
-    pub fn sceRegCloseCategory(dir_handle: Handle) -> i32;
+    pub fn sceRegCloseCategory(dir_handle: RegistryHandle) -> i32;
 
     #[psp(0x0D69BF40)]
     /// Flush the registry directory to disk
@@ -141,7 +141,7 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 on success, < 0 on error
-    pub fn sceRegFlushCategory(dir_handle: Handle) -> i32;
+    pub fn sceRegFlushCategory(dir_handle: RegistryHandle) -> i32;
 
     #[psp(0xD4475AA8, i5)]
     /// Get a key's information
@@ -158,9 +158,9 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegGetKeyInfo(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         name: *const u8,
-        key_handle: *mut Handle,
+        key_handle: *mut RegistryHandle,
         type_: *mut KeyType,
         size: *mut usize,
     ) -> i32;
@@ -179,7 +179,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegGetKeyInfoByName(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         name: *const u8,
         type_: *mut KeyType,
         size: *mut usize,
@@ -199,8 +199,8 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegGetKeyValue(
-        dir_handle: Handle,
-        key_handle: Handle,
+        dir_handle: RegistryHandle,
+        key_handle: RegistryHandle,
         buf: *mut c_void,
         size: usize,
     ) -> i32;
@@ -219,7 +219,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegGetKeyValueByName(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         name: *const u8,
         buf: *mut c_void,
         size: usize,
@@ -239,7 +239,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegSetKeyValue(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         name: *const u8,
         buf: *const c_void,
         size: usize,
@@ -257,7 +257,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegGetKeysNum(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         num: *mut i32,
     ) -> i32;
 
@@ -274,7 +274,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegGetKeys(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         buf: *mut u8,
         num: i32,
     ) -> i32;
@@ -293,7 +293,7 @@ psp_extern! {
     ///
     /// 0 on success, < 0 on error
     pub fn sceRegCreateKey(
-        dir_handle: Handle,
+        dir_handle: RegistryHandle,
         name: *const u8,
         type_: i32,
         size: usize,
@@ -309,5 +309,5 @@ psp_extern! {
     /// # Return Value
     ///
     /// 0 on success, < 0 on error
-    pub fn sceRegRemoveRegistry(key: *mut Key) -> i32;
+    pub fn sceRegRemoveRegistry(key: *mut RegistryKey) -> i32;
 }

